pymc3/distributions/mixture.py

```diff
@@ -108,86 +130,220 @@ def comp_dists(self):
+                generator = Mixture._comp_dist_random_wrapper(comp_dist.random)
+                self._generators.append(generator)
+            self.is_multidim_comp = False
+
+    @staticmethod
+    def _comp_dist_random_wrapper(random):
+        """Wrap the comp_dists.random method to take the kwarg raw_size_ and
+        use it's value to replace the size parameter. This is needed because
+        generate_samples makes the size value compatible with the
+        scipy.stats.*.rvs, where size has a different meaning than in the
+        distributions' random methods.
+        """
+        def wrapped_random(*args, **kwargs):
+            raw_size_ = kwargs.pop('raw_size_', None)
+            # Distribution.random's signature is always (point=None, size=None)
+            # so size could be the second arg or be given as a kwarg
+            if len(args) > 1:
+                args[1] = raw_size_
             else:
-                try:
-                    self.comp_dists.random(size=23)
-                    self._comp_dists_vect = True
-                except Exception:
-                    # The comp_dists cannot call random with non None size or
-                    # without knowledge of the point so we assume that we will
-                    # have to iterate calls to random to get the correct size
-                    self._comp_dists_vect = False
+                kwargs['size'] = raw_size_
+            return random(*args, **kwargs)
+        return wrapped_random
 
     def _comp_logp(self, value):
         comp_dists = self.comp_dists
 
-        try:
-            value_ = value if value.ndim > 1 else tt.shape_padright(value)
-
+        if self.is_multidim_comp:
+            # Value can be many things. It can be the self tensor, the mode
+            # test point or it can be observed data. The latter case requires
+            # careful handling of shape, as the observed's shape could look
+            # like (repetitions,) + dist_shape, which does not include the last
+            # mixture axis. For this reason, we try to eval the value.shape,
+            # compare it with self.shape and shape_padright if we infer that
+            # the value holds observed data
+            try:
+                val_shape = tuple(value.shape.eval())
+            except AttributeError:
+                val_shape = value.shape
+            except theano.gof.MissingInputError:
+                val_shape = None
+            ndim = value.ndim
+            if val_shape is not None:
+                # If value does not hold observed data, so we can use its ndim
+                # safely to determine shape padding
+                try:
+                    self_shape = tuple(self.shape)
+                except AttributeError:
+                    # Happens in __init__ when computing self.logp(comp_modes)
+                    self_shape = None
+                comp_shape = tuple(comp_dists.shape)
+                if (
```

Comment:
There is no else clause here? I think you should raise an exception

Reply:
Oops! You're right! Thanks for catching that. I'll add an else clause, or restructure the conditions a bit.